ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/CanHardFork.hs

```diff
@@ -595,3 +625,82 @@ translateValidatedTxMaryToAlonzoWrapper ::
        (ShelleyBlock (TPraos c) (AlonzoEra c))
 translateValidatedTxMaryToAlonzoWrapper ctxt = InjectValidatedTx $
     fmap unComp . eitherToMaybe . runExcept . SL.translateEra ctxt . Comp
+
+{-------------------------------------------------------------------------------
+  Translation from Alonzo to Babbage
+-------------------------------------------------------------------------------}
+
+translateLedgerStateAlonzoToBabbageWrapper ::
+     (Praos.PraosCrypto c, TPraos.PraosCrypto c)
+  => RequiringBoth
+       WrapLedgerConfig
+       (Translate LedgerState)
+       (ShelleyBlock (TPraos c) (AlonzoEra c))
+       (ShelleyBlock (Praos c) (BabbageEra c))
+translateLedgerStateAlonzoToBabbageWrapper =
+    RequireBoth $ \_cfgAlonzo cfgBabbage ->
+      Translate $ \_epochNo ->
+        unComp . SL.translateEra' (getBabbageTranslationContext cfgBabbage) . Comp . transPraosLS
+  where
+    transPraosLS ::
+      LedgerState (ShelleyBlock (TPraos c) (AlonzoEra c)) ->
+      LedgerState (ShelleyBlock (Praos c) (AlonzoEra c))
+    transPraosLS (ShelleyLedgerState wo nes st) =
+      ShelleyLedgerState
+        { shelleyLedgerTip = case wo of
```

Comment:
Maybe we should define `castShelleyTip` so that we can replace this with `fmap castShelleyTip`?

Reply:
I pushed up a commit for this on my `nfrisby/praos-protocol` variant branch.

Reply:
Clarke cherry-picked them, works 👍